src/ZODB/tests/racetest.py

```diff
@@ -0,0 +1,515 @@
+                # zconn cache
+                _state_invalidate_half1(root)
+
+                try:
+                    spec.assertStateOK(root)
+                except AssertionError as e:
+                    msg = "T%s: %s\n" % (tx, e)
+                    msg += _state_details(root)
+                    failure[tx] = msg
+                    failed.set()
+
+                # change objects once in a while
+                if randint(0, 4) == 0:
+                    # print("T%s: modify" % tx)
+                    spec.next(root)
+                    spec.assertStateOK(root)
+
+                try:
+                    transaction.commit()
+                except POSException.ConflictError:
+                    # print('conflict -> ignore')
+                    transaction.abort()
+
+                zconn.close()
+
+            try:
+                for i in range(N):
+                    # print('T%s.%d' % (tx, i))
+                    t_()
+                    if failed.is_set():
+                        break
+            except:  # noqa: E722 do not use bare 'except'
+                failed.set()
+                raise
+            finally:
+                db.close()
+
+        # run the workers concurrently.
+        init()
+
+        N = 100
+        tg = []
+        for x in range(nwork):
+            t = threading.Thread(name='T%d' % x, target=T, args=(x, N))
+            t.start()
+            tg.append(t)
+
+        for t in tg:
+            t.join(60)
+
+        if failed.is_set():
+            self.fail('\n\n'.join([_ for _ in failure if _]))
+
+    # verify storage for race in between client disconnect and external
+    # invalidations. https://github.com/zopefoundation/ZEO/issues/209
+    #
+    # This test is simlar to check_race_load_vs_external_invalidate, but
+    # increases the number of workers and also makes every worker to repeatedly
+    # reconnect to the storage, so that the probability of disconection is
+    # high. It also uses T2ObjectsInc2Phase instead of T2ObjectsInc because if
+    # an invalidation is skipped due to the disconnect/invalidation race,
+    # T2ObjectsInc won't catch the bug as both objects will be either in old
+    # state, or in new state after the next transaction. Contrary to that, with
+    # T2ObjectsInc2Phase the invariant will be detected to be broken on the
+    # next transaction.
+    def check_race_external_invalidate_vs_disconnect(self):
+        return self._check_race_xxx_vs_external_disconnect(
+                                                T2ObjectsInc2Phase())
+
+    @with_high_concurrency
+    def _check_race_xxx_vs_external_disconnect(self, spec):
+        assert isinstance(spec, ISpec)
+
+        # init initializes the database according to the spec.
+        def init():
+            db = self.dbopen()
+            _state_init(db, spec)
+            db.close()
+
+        nwork = 8*8   # nwork^2 from _check_race_load_vs_external_invalidate
```

Comment:
@d-maurer, thanks for feedback.

Yes, this test is longer, but I've tried to keep its running time limited: on my system `check_race_loadopen_vs_local_invalidate` takes ~ 0.8s, `check_race_load_vs_external_invalidate` takes ~ 0.7s, while hereby `check_race_external_invalidate_vs_disconnect` takes ~ 7 seconds. Yes, the test is run several time for all adaptations variants, but still, even with 7 seconds of runtime, it does not fail 100% for a particular run. But it fails with close to 100% probability in on f the runs for all adaptation kinds. I belive we can accept those extra 7*4 seconds to be run by github automatically.

Still, during development, this time overhead might be indeed inconvenient. So I've tried to address this in https://github.com/zopefoundation/ZODB/commit/2a2924a53884e0a5a44afaa013c57eefa76c0553. Hope it is ok.


P.S. I will try to address your other input incrementally as time permits.

Reply:
( fixed the last patch to make lint happy. )
<details>
<summary>interdiff</summary>

```diff
--- a/src/ZODB/tests/util.py
+++ b/src/ZODB/tests/util.py
@@ -113,7 +113,8 @@ def setUp(self):
 
     tearDown = tearDown
 
-    # propagate .level from tested method to TestCase so that e.g. @long_test works
+    # propagate .level from tested method to TestCase so that e.g. @long_test
+    # works
     @property
     def level(self):
```

</details>